text/0000-rand-crate-redesign.md

```diff
@@ -0,0 +1,991 @@
+deterministic algorithm for generating *random-like* numbers. Such algorithms
+typically have a fixed-size state space, a *seed* function to produce an initial
+state from some value, an *advance* function to step from one state to the next,
+and a *generate* function to yield a value in the output domain (type). This
+implies the following properties:
+
+*   Generated numbers are reproducible, when the algorithm and seed (or initial
+    state) is known
+*   All PRNGs have a finite period: eventually the *advance* function will
+    reproduce a prior state (not necessarily the initial one), and the sequence
+    of numbers will repeat
+
+Given a fixed state-space of `k` bits, at most `2^k` states are possible. A
+good PRNG should have a large period, possibly close to `2^k`; this period
+may depend on the seed value, but good PRNGs should ensure a large period for
+all seeds.
+
+To be useful, a PRNGs should usually also have the following properties:
+
+*   Require little memory and have fast computation
+*   Generated values are uniformly distributed across the output domain
+*   The distribution of each value is indepedent of previously generated
+    values(1)
+*   For cryptographic applications, it is not possible to predict knowledge of
+    prior values does not aid in predicting the next value(1)
+
+Note (1): obviously once a PRNG has completed its initial period and started
+repeating itself, the above properties are no longer possible. It is therefore
+required that the period is very long.
+
+Further, a PRNG may offer a *jump-ahead* function to quickly calculate the
+state after a large arbitrary number of steps `v`. This allows a random number
+stream to be deterministically partitioned into a number of sub-streams.
+
+L'Ecuyer provides some more background on PRNGs in
+[Random Number Generation](https://scholar.google.com/citations?view_op=view_citation&hl=en&user=gDxovowAAAAJ&citation_for_view=gDxovowAAAAJ:d1gkVwhDpl0C) (Springer, 2012).
+
+## Guide-level explanation
+
+Since this concerns one (or more) crates outside the standard library, it is
+assumed that these crates should be self-documenting. Much of this documentation
+still needs writing, but must wait on design decisions.
+
+In my opinion, [the extensive examples](https://docs.rs/rand/0.3.16/rand/#examples)
+in the crate API documentation should be moved to a book or example project.
+They are well written, but do not really belong in API documentation.
+
+# Reference-level explanation
+
+## Generation API
+
+This section concerns the `Rng` trait and extension traits, but not
+specifically implementations or generation of values of other types.
+
+The `Rng` trait covers *generation* of values. It is not specific to
+[deterministic] PRNGs; instead there is an extension trait `SeedableRng: Rng`
+covering deterministic seeding.
+
+It has been proposed to rename `Rng` to `Generator`; this proposal has not
```

Comment:
Shortness itself isn't the best virtue. This isn't the 1980s any more, we can stand to waste entire characters of memory in our source files. Consider how often _outside of the rand crate_ you'll be writing down all these types. Probably into a struct definition or two and then (if we do it right and go a "batteries included" route) very rarely besides that.

Which isn't to say that shortness is necessarily bad either. It's just totally neutral compared to overall clarity. We could still call it `Gen` if we do like shortness that much. Examples:

* `Gen`
* `Gen: SeedableGen`
* `ReseedingGen<G:SeedableGen>`
* etc